### DISCUSSION TITLE
When a protocol decoders sees an error, safe only that data, and discard the rest during a very long capture (days, weeks)

### DISCUSSION THREAD

**Discussion by cdwijs** [gusmanb/logicanalyzer#123] [Ideas]
One situation where I need a logic analyser, is when I have made firmware for two micro controllers that talk to each other via RS232. Sometimes one of the micro controllers makes a mistake. This then results in the other side not answering.

I would like to let Logic Analyser run overnight, and capture all the serial data, and some other signals from both PCB's containing the microcontrollers. The problem is that the resulting capture will be VERY large, and I only need the data around the time it went wrong.

Now i would like to see that the serial decoder is able to see a missing response pn the RX line to a request on the TX line. When that happens, i would like to have a file that contains all the signals from 1 second before the serial TX, to 1 second after the timeout. 

When i come back the next day, i then have a series of capture files, each one containing a few seconds of data, around a serial time out condition.

Bonus points if this can be generic, so a user can set triggers if a particulair serial string is received, if there's a serial frame error, if there's a crc error in a MODBUS message, If an i2c sensor sets a certain bit of a certain register, and so on. Maybe even with some logic, for instance an OR condition.

**Comment by gusmanb** [comment]
Sorry but this project will never be able to do this. It is impossible to process the data in realtime. I'm pushing the RP2040 beyond its limits, the target of the project is to have a really cheap logic analyzer that is able to capture signals at a rate up to 100Mhz, for that there are compromises made. I can only capture data directly to the memory of the pico, is impossible to transfer the data directly to the PC (the USB of the pico is too slow) so that functionality cannot be implemented.

The best you can do is to use the burst mode to only capture data when an event happens, but for the case of a timeout you will already have lost the data.

**Reply by cdwijs** [reply]
Would you be willing to consider a continuous sampling mode like the sigrok-pico [*] is doing? With continuous sampling, all the proposed processing can be done at the PC side.

[*] https://github.com/pico-coder/sigrok-pico

**Reply by gusmanb** [reply]
The problem with the continuous sampling mode is the speed, the pico USB speed is very limited, the best you can get is around 100-120Kb/s, that really limits the possibilities, in the best case you could sample 8 channels at 100Khz, that's mostly useless...

**Reply by cdwijs** [reply]
Sigrok-pico archives a streaming rate of 120Msamp / sec when only digital channels are used:
https://github.com/pico-coder/sigrok-pico/blob/main/PICO_Sigrok_Rates.PNG

**Reply by gusmanb** [reply]
> Sigrok-pico archives a streaming rate of 120Msamp / sec when only digital channels are used: https://github.com/pico-coder/sigrok-pico/blob/main/PICO_Sigrok_Rates.PNG

Sigrok-pico does NOT achieve a streaming rate of 120Ms, read the table you have posted, it reaches the 120Msps only when you sample less than n samples, basically what fits in memory, once it's filled it goes down dramatically to 160Ksp/s, more or less what I stated.
